pkg/repositories/transformers/task_execution.go

```diff
@@ -240,6 +242,94 @@ func mergeCustom(existing, latest *_struct.Struct) (*_struct.Struct, error) {
 	return &response, nil
 }
 
+// mergeExternalResource combines the lastest ExternalResourceInfo proto with an existing instance
+// by updating fields and merging logs.
+func mergeExternalResource(existing, latest *event.ExternalResourceInfo) *event.ExternalResourceInfo {
+	if existing == nil {
+		return latest
+	}
+
+	if latest == nil {
+		return existing
+	}
+
+	if latest.ExternalId != "" && existing.ExternalId != latest.ExternalId {
+		existing.ExternalId = latest.ExternalId
+	}
+	// we only update if the index is equal so updating existing.Index is not necessary
+	if latest.RetryAttempt != 0 && existing.RetryAttempt != latest.RetryAttempt {
+		existing.RetryAttempt = latest.RetryAttempt
+	}
+	existing.Phase = latest.Phase
+	if latest.CacheStatus != core.CatalogCacheStatus_CACHE_DISABLED && existing.CacheStatus != latest.CacheStatus {
+		existing.CacheStatus = latest.CacheStatus
+	}
+	existing.Logs = mergeLogs(existing.Logs, latest.Logs)
+
+	return existing
+}
+
+// mergeExternalResources combines lists of external resources. This involves appending new
+// resources and updating in-place resources attributes.
+func mergeExternalResources(existing, latest []*event.ExternalResourceInfo) []*event.ExternalResourceInfo {
+	if len(latest) == 0 {
+		return existing
+	}
+
+	for i, externalResource := range latest {
+		var index int
+		if externalResource.GetIndex() == 0 && externalResource.GetRetryAttempt() == 0 && externalResource.GetPhase() == 0 {
+			// updating an external resource to retryAttempt 0 and phase 0 is invalid, because that
+			// is the initial state, so we use this condition to signify an external resource
+			// without these parameters (ex. a simple ExternalResourceID).
+			index = i
```

Comment:
is there any chance another externalResource in latest might have already claimed index `i`? could you leave a comment for this method about the general assumption behind indices, like any ordering constraints for example?

Reply:
I updated the comment. I think it makes sense, please let me know what you think.

Reply:
thank you!